Refactor Core Classes to Improve Readability, Reusability, and Maintainability

DishaPatel126 opened the pull request:
## Pull Request: Refactoring for Maintainability and Clarity

This pull request introduces several refactorings across key components of the `nats.java` codebase.  
The goal is to **enhance clarity**, **reduce code duplication**, and promote **maintainable, modular code** without altering existing functionality.

The changes are grouped into **two sets of well-known refactoring patterns**:

---

### 🔹 Set I

#### 🔸 Rename Method
- **File**: `FileAuthHandler.java`  
- **Change**: Renamed `readKeyChars()` → `extractKeyFromFile()` for better clarity of purpose.

#### 🔸 Extract Method
- **File**: `ErrorListenerConsoleImpl.java`  
- **Change**: Extracted logging logic into a new method to improve readability and adhere to the single responsibility principle.

#### 🔸 Introduce Explaining Variable
- **File**: `KeyValueEntry.java`  
- **Change**: Refactored `hashCode()` method by introducing meaningful intermediate variables to improve comprehension.

---

### 🔹 Set II

#### 🔸 Extract Class
- **File**: `Encoding.java`  
- **Change**: Extracted independent utilities into new classes:
  - `Base64Utils.java`
  - `Base32Utils.java`
  - `JsonUtilsLite.java`
  - `UriUtilsLite.java`

#### 🔸 Pull-Up Method
- **Files**:
  - **Superclass**: `NatsDispatcher.java`  
  - **Subclass**: `NatsDispatcherWithExecutor.java`  
- **Change**: Moved `handleMessage(...)` from subclass to superclass. It is now overridden only where asynchronous execution is required.

#### 🔸 Replace Conditional with Polymorphism
- **File**: `MessageInfo.java`  
- **Change**: Refactored complex constructor logic by delegating parsing to dedicated private methods:
  - `parseFromStatus()`
  - `parseFromDirectMessage()`
  - `parseFromJson()`  
  This improves readability, modularity, and testability.

scottf commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.8.3-bin.zip
```
Not changing gradle. All gradle changes must be removed

scottf commented on src/main/java/io/nats/client/api/KeyValueEntry.java:
```diff
@@ -136,7 +137,7 @@ public boolean equals(Object o) {
     public int hashCode() {
         int result = bucketAndKey.hashCode();
-        result = 31 * result + Long.hashCode(revision);
+        result = PRIME * result + Long.hashCode(revision);
```
This is generated code and not necessary to change.

scottf commented on src/main/java/io/nats/client/api/MessageInfo.java:
```diff
@@ -39,6 +39,18 @@ public class MessageInfo extends ApiResponse<MessageInfo> {
+        long numPending;
+        Status status;
+    }
+
```
Not necessary.

scottf commented on src/main/java/io/nats/client/impl/ErrorListenerConsoleImpl.java:
```diff
@@ -23,31 +23,31 @@ public class ErrorListenerConsoleImpl implements ErrorListener {
     @Override
     public void errorOccurred(final Connection conn, final String error) {
-        System.out.println(supplyMessage("[SEVERE] errorOccurred", conn, null, null, "Error: ", error));
+        log("[SEVERE]", "errorOccurred", conn, null, null, "Error: ", error);
```
No. 
1. This is perfectly fine. A user can always make their own.
2. This would actually be a breaking change. Someone could rely out the specific text that this class outputs to a log and that would break their system.

scottf commented on src/main/java/io/nats/client/impl/FileAuthHandler.java:
```diff
@@ -108,7 +108,7 @@ private char[] extract(CharBuffer data, int headers) {
     }
 
-    private char[] readKeyChars() throws IOException {
+    private char[] extractKeyFromFile() throws IOException {
```
This is not necessary, but would be fine.

scottf commented on src/main/java/io/nats/client/impl/NatsDispatcher.java:
```diff
@@ -132,6 +132,11 @@ public void run() {
+    protected void handleMessage(NatsSubscription sub, NatsMessage msg, MessageHandler handler) throws InterruptedException {
+        handler.onMessage(msg);
+    }
+
```
Where is this used?

scottf reviewed (commented):
We greatly appreciate contributions, but this seems much like a contribution designed simply to get a mark that says I contributed to open source. It is just not necessary to refactor for refactoring sake.
If there is a legitimate bug fix or feature/improvement, then refactoring in areas surrounding that change are welcome.

DishaPatel126 commented:
Greetings Scott Fauerbach,

Thank you for taking the time to review my recent pull request to the
nats.java repository.

I’d like to offer some context behind the changes I submitted. This was
part of an academic assignment focused on code refactoring. The changes
weren’t arbitrary — I used a tool to detect design and implementation
smells, and addressed only those findings through standard refactoring
techniques like method renaming and class extraction. This was done with
the intention of improving readability and modularity in targeted areas. My
intent was never to refactor for the sake of it or to merely "check a box"
for open source contribution, but rather a structured exercise grounded in
analysis and software engineering principles.

Additionally, regarding the Gradle configuration: I updated the Gradle
version only because the original version failed to work in my local
development setup. I understand the importance of consistency in build
tooling and I respect your decision and feedback fully.

I’m grateful for the opportunity to engage with a production-grade open
source project — it's been a valuable learning experience.

Regards,
Disha Patel


On Sat, 29 Mar 2025 at 10:57, Scott Fauerbach ***@***.***>
wrote:

> Closed #1297 <https://github.com/nats-io/nats.java/pull/1297>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nats-io/nats.java/pull/1297#event-17043230353>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A33TICRV3DXBOVKMS66R3G32W2RDRAVCNFSM6AAAAAB2A4MJPWVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJXGA2DGMRTGAZTKMY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

scottf commented:
Also...
* Commits need to be verified.
* All new source files must have the copyright.

scottf commented:
@DishaPatel126 Sent direct email.
